p2p/src/net/mock/transport/impls/stream_adapter/wrapped_transport/wrapped_listener.rs

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use async_trait::async_trait;
+use futures::future::BoxFuture;
+
+use crate::{
+    net::mock::{
+        peer::Role,
+        transport::{
+            impls::stream_adapter::{
+                traits::StreamAdapter, wrapped_transport::utils::HandshakeFut,
+            },
+            TransportListener, TransportSocket,
+        },
+    },
+    Result,
+};
+
+/// A listener object that handles new incoming connections, and does any required hand-shakes (see members' comments)
+pub struct AdaptedListener<S: StreamAdapter<T::Stream>, T: TransportSocket> {
+    pub stream_adapter: S,
+    pub listener: T::Listener,
+    #[allow(clippy::type_complexity)]
+    pub handshakes: Vec<(BoxFuture<'static, Result<S::Stream>>, T::Address)>,
+}
+
+#[async_trait]
+impl<S: StreamAdapter<T::Stream>, T: TransportSocket> TransportListener<S::Stream, T::Address>
+    for AdaptedListener<S, T>
+{
+    async fn accept(&mut self) -> Result<(S::Stream, T::Address)> {
+        loop {
+            tokio::select! {
+                handshake = HandshakeFut::<S, T>(&mut self.handshakes) => {
+                    return Ok(handshake);
+                }
+                accept_res = self.listener.accept() => {
+                    match accept_res {
+                        Ok((base, addr)) => {
+                            // Store active handshakes because accept must be cancel safe
+                            let handshake = self.stream_adapter.handshake(base, Role::Inbound);
+                            self.handshakes.push((handshake, addr));
```

Comment:
Maybe I'm missing something... but seems like this vector will be growing forever? Shouldn't this be some kind of double-ended queue, where handshakes that are done are popped out?

Reply:
No, it won't be growing forever because we remove complete handshakes here: https://github.com/mintlayer/mintlayer-core/blob/70222be5398046ebeb80258be13a2602cfe0dc98/p2p/src/net/mock/transport/impls/stream_adapter/wrapped_transport/utils.rs#L39
Double-ended queue won't help because handshakes might complete in random order.
Though it might be a good idea to bound total count of active handshakes.

Reply:
Alrighty. I'll comment there for follow-up.